### ISSUE TITLE
Compilation error in UTF8-CPP when compiled using g++ 5.4 or onwards

### ISSUE THREAD

**Issue by waghamolm** [nemtrif/utfcpp#16]
Hello @nemtrif ,

I have [main.cc.zip](https://github.com/nemtrif/utfcpp/files/2035377/main.cc.zip) with some code snippet, which uses library functionality.

When that code is compiled on Ubuntu 16.04 with g++ 5.4, it gives compilation error. See 
![utf8_compilation_error_on_gcc_5_4](https://user-images.githubusercontent.com/18116300/40485258-ab909500-5f7b-11e8-8426-4aaf585f11ac.png) for more details.

In order to resolve the compilation issue, we need to change class utf8::iterator declaration in 'checked.h' as given below.
We need to include "cstddef" for std::ptrdiff_t
`
    template <typename octet_iterator>
    class iterator : public std::iterator <
            std::bidirectional_iterator_tag,
            uint32_t,
            std::ptrdiff_t,
            const uint32_t*,
            const uint32_t& >
`
But, even after resolving the compilation issue, some more changes may be needed in the library code. Since, output of the program is incorrect.
Expected Result :
Within String ℇ¥ǢǾ¥ǽƱ
Find String ¥
String found at position 2

Actual Result :
Within String ℇ¥ǢǾ¥ǽƱ
Find String ¥
Error occurred

Please note that, even after changes in 'checked.h', we get 'expected result' when compiled with g++ 4.8
To summarize, it looks like code behaves incorrectly with g++ version 5.4.

**Comment by nemtrif** [comment]
Works fine with gcc 7.3.0
